#!/usr/bin/env python3
#author:   yangm@idsse.ac.cn
#versions: 0.0.1
#versions: 0.0.2 2023-07-26 22:42
#versions: 0.0.5 2026-01-15 00:24

import os
import sys
import re
import argparse
import pandas as pd
import linkTab

# --------------------------------------------------
# merge_vctg_info.py is appliable to merge all viral information generated by VirSorter2, VIBERANT, geNomad, and DeepVirFinder.
# If you need to add a certain new virus sequence identification software, the following revisation must be performed:
# 1. Add a new element or key-value pairs for each dictionary before the first function;
# 2. Specify the relative path of the result file and provirus coordinate file in the 'resultFile' function based on the ResultDict naming conventions to match the actual file paths;
# 3. Define a new function named "handle_<new_tool_name>" before the "TOOL_HANDLER" dictionary to extract the key viral information generated by the new tool;
# 4. Add a new line in the calcCtgScore function to convert scores for the new software.
# --------------------------------------------------

# --------------------------------------------------
# Define variables for the tools and their result information
# --------------------------------------------------

Tools = ['virsorter2', 'vibrant', 'genomad', 'deepvirfinder']

TOOL_ALIAS = {
    'vs2': 'virsorter2',
    'vb': 'vibrant',
    'gn': 'genomad',
    'dvf': 'deepvirfinder'
}

ResultDict = {
    'virsorter2': '{0}/vs2-pass1/final-viral-score.tsv,{0}/vs2-pass1/final-viral-boundary.tsv',
    'vibrant': '{0}/VIBRANT_{1}/VIBRANT_results_{1}/VIBRANT_genome_quality_{1}.tsv,{0}/VIBRANT_{1}/VIBRANT_results_{1}/VIBRANT_integrated_prophage_coordinates_{1}.tsv',
    'genomad': '{0}/genomad/{1}_summary/{1}_virus_summary.tsv',
    'deepvirfinder': '{0}/deepvirfinder'
}

CsvDict = {
    'virsorter2': '{}/vs2_viral_info.tsv',
    'vibrant': '{}/vb_viral_info.tsv',
    'deepvirfinder': '{}/dvf_viral_info.tsv',
    'genomad': '{}/gn_viral_info.tsv'
}

ColsDict = {
    'virsorter2': {
        'dsDNAphage': 'vs2_dsDNAphage',
        'ssDNA': 'vs2_ssDNA',
        'NCLDV': 'vs2_NCLDV',
        'RNA': 'vs2_RNA',
        'lavidaviridae': 'vs2_lavidaviridae',
        'max_score': 'vs2_max_score',
        'max_score_group': 'vs2_max_score_group',
        'hallmark': 'vs2_hallmark',
        'viral': 'vs2_viral',
        'cellular': 'vs2_cellular',
        'vs2_partial': 'vs2_partial'
    },
    'vibrant': {
        'type': 'vb_lifestyle',
        'vb_isPhage': 'vb_isPhage',
        'vb_partial': 'vb_partial'
    },
    'deepvirfinder': {
        'len': 'length',
        'score': 'dvf_v_score',
        'pvalue': 'dvf_pvalue'
    },
    'genomad': {
        'topology': 'gn_topology',
        'coordinates': 'gn_coordinates',
        'n_genes': 'gn_genes',
        'genetic_code': 'gn_genetic_code',
        'virus_score': 'gn_v_score',
        'fdr': 'gn_fdr',
        'n_hallmarks': 'gn_hallmarks',
        'marker_enrichment': 'gn_marker_enrichment',
        'taxonomy': 'gn_taxonomy',
        'gn_partial': 'gn_partial'
    }
}

def parse_merge_tools(m_str):
    tools = []
    for t in m_str.split('-'):
        if t not in TOOL_ALIAS:
            raise ValueError(f'Unknown tool alias: {t}')
        tools.append(TOOL_ALIAS[t])
    return list(dict.fromkeys(tools))

# --------------------------------------------------
# Get the real path of a certain tool' result or coordinate file  
# --------------------------------------------------


def resultFile(name, tool, wkdir):
    wkdir = wkdir.rstrip('/')
    result = ResultDict[tool] 
    if tool == 'virsorter2':
        result = result.format(wkdir)
    elif tool == 'deepvirfinder':
        result = result.format(wkdir)
        file_name = os.listdir(result)[0]
        result += '/' + file_name
    elif tool == 'vibrant' or tool == 'genomad':
        result = result.format(wkdir, name)
    return result

# --------------------------------------------------
# Tool handlers (only discussed modifications applied)
# --------------------------------------------------

def handle_virsorter2(result):
    score_file, boundary_file = result.split(',')
    df_s = pd.read_csv(score_file, sep='\t', header=0)
    df_s[['Contig', 'vs2_region']] = df_s['seqname'].str.split(r'\|\|', expand=True)
    if 'length' in df_s.columns:
        df_s = df_s.drop(columns=['length'])
    score_agg = {
        'vs2_region': lambda x: ','.join(sorted(set(x))), 
        'dsDNAphage': 'max', 
        'ssDNA': 'max',
        'NCLDV': 'max',
        'RNA': 'max',
        'lavidaviridae': 'max',
        'max_score': 'max',
        'max_score_group': 'first',
        'hallmark': 'max',
        'viral': 'max',
        'cellular': 'max'
    }
    df_score = df_s.groupby('Contig', as_index=False).agg(score_agg)
    try:
        df_b = pd.read_csv(boundary_file, sep='\t', header=0)
        df_b[['Contig', 'vs2_region']] = df_b['seqname_new'].str.split(r'\|\|', expand=True)
        df_b['vs2_nucl_range'] = df_b['trim_bp_start'].astype(str) + '-' + df_b['trim_bp_end'].astype(str)
        df_b['vs2_orf_range'] = df_b['trim_orf_index_start'].astype(str) + '-' + df_b['trim_orf_index_end'].astype(str)
        df_boundary = df_b.groupby('Contig', as_index=False).agg({'vs2_region': lambda x: ','.join(sorted(set(x))), 'vs2_nucl_range': lambda x: ','.join(x), 'vs2_orf_range': lambda x: ','.join(x)})
    except (FileNotFoundError, pd.errors.EmptyDataError, ValueError, KeyError):
        df_boundary = pd.DataFrame(columns=['Contig', 'vs2_region', 'vs2_nucl_range', 'vs2_orf_range'])
    df = df_score.merge(df_boundary, on='Contig', how='left', suffixes=('', '_boundary'))
    if 'vs2_region_boundary' in df.columns:
        df['vs2_region'] = df['vs2_region_boundary'].fillna(df['vs2_region'])
        df = df.drop(columns=['vs2_region_boundary'])
    df['vs2_nucl_range'] = df['vs2_nucl_range'].fillna('')
    df['vs2_orf_range'] = df['vs2_orf_range'].fillna('')
    return df

def handle_vibrant(result):
    quality_file, provirus_range_file = result.split(',')
    df_q = pd.read_csv(quality_file, sep='\t', header=0)
    df_q = df_q[['scaffold', 'type']].drop_duplicates()
    tmp = df_q['scaffold'].str.split(r'_fragment_', expand=True)
    df_q['Contig'] = tmp[0]
    if tmp.shape[1] > 1:
        df_q['vb_partial'] = tmp[1].apply(lambda x: f'fragment_{x}' if pd.notna(x) else '')
    else:
        df_q['vb_partial'] = ''
    df_q = df_q[['Contig', 'type', 'vb_partial']]
    try:
        df_p = pd.read_csv(provirus_range_file, sep='\t', usecols=['scaffold', 'fragment', 'nucleotide start', 'nucleotide stop', 'protein start', 'protein stop'])
        df_p = df_p.rename(columns={'scaffold': 'Contig'})
        df_p['vb_partial'] = df_p['fragment'].str.extract(r'(fragment_\d+)')
        df_p['vb_nucl_range'] = df_p['nucleotide start'].astype(str) + '-' + df_p['nucleotide stop'].astype(str)
        df_p['vb_prot_range'] = df_p['protein start'].astype(str) + '-' + df_p['protein stop'].astype(str)
        prophage_agg = {
            'vb_partial': lambda x: ','.join(sorted(set(x))), 
            'vb_nucl_range': lambda x: ','.join(x), 
            'vb_prot_range': lambda x: ','.join(x)
        }
        df_p_agg = df_p.groupby('Contig', as_index=False).agg(prophage_agg)
    except (FileNotFoundError, pd.errors.EmptyDataError, ValueError):
        df_p_agg = pd.DataFrame(columns=['Contig', 'vb_partial', 'vb_nucl_range', 'vb_prot_range'])
    df = df_q.groupby('Contig', as_index=False).agg({'type': 'first'}).merge(df_p_agg, on='Contig', how='left')
    df['vb_partial'] = df['vb_partial'].fillna('')
    df['vb_nucl_range'] = df['vb_nucl_range'].fillna('')
    df['vb_prot_range'] = df['vb_prot_range'].fillna('')
    df['vb_isPhage'] = 1
    return df

def handle_genomad(result):
    df = pd.read_csv(result, sep='\t', header=0)
    try:
        df[['Contig', 'gn_partial']] = df['seq_name'].str.split(r'\|pro', expand=True)
        df['gn_partial'] = 'pro' + df['gn_partial']
    except ValueError:
        df['Contig'] = df['seq_name']
        df['gn_partial'] = ''
    df = df.drop(columns=['length'])
    agg_dict = {
        'topology': 'first',
        'coordinates': lambda x: ','.join(x.dropna().astype(str)),
        'gn_partial': lambda x: ','.join(x.dropna().astype(str)),
        'n_genes': lambda x: x.dropna().astype(int).sum(),
        'genetic_code': 'first',
        'n_hallmarks': lambda x: x.dropna().astype(int).max(),
        'marker_enrichment': lambda x: x.dropna().astype(float).max(),
        'taxonomy': lambda x: ','.join(x.dropna().astype(str)),
        'virus_score': 'max',
        'fdr': 'min'
    }
    return df.groupby('Contig', as_index=False).agg(agg_dict)

def handle_dvf(result):
    df = pd.read_csv(result, sep='\t', header=0)
    df = df.drop(columns=['len'])
    return df.rename(columns={'name': 'Contig'})

TOOL_HANDLER = {
    'virsorter2': handle_virsorter2,
    'vibrant': handle_vibrant,
    'genomad': handle_genomad,
    'deepvirfinder': handle_dvf
}

# --------------------------------------------------

def listToFile(list_l, list_f):
    LIST = open(list_f, 'w')
    for ctg in list_l:
        LIST.write(f'{ctg}\n')
    return 0

#vCtgMerge: Integrate the contig lists identified by all selected viral detection tools, and output the individual contig list for each tool in parallel.
def vCtgMerge(name, wkdir, tools=['gn']):
    all_ctgs = []
    for tool in tools:
        result = resultFile(name, tool, wkdir)
        df = TOOL_HANDLER[tool](result)
        all_ctgs.extend(df['Contig'].tolist())
        all_ctgs = list(set(all_ctgs))
        df = df.rename(columns=ColsDict.get(tool, {}))
        csv_name = CsvDict[tool].format(wkdir)
        df.to_csv(csv_name, index=False, sep='\t')
    return all_ctgs

def calcCtgScore(all_merged_ctg_tsv, tools=['gn'], filt_mode='permissive'):
    df = pd.read_csv(all_merged_ctg_tsv, sep='\t')
    score = 0
    if 'virsorter2' in tools:
        df['vs2_score'] = df.apply(lambda x : 2 if x['vs2_max_score'] >= 0.9 else (1 if x['vs2_max_score'] >= 0.7 and x['vs2_hallmark'] >= 1 else 0), axis = 1)
        score += df['vs2_score']
    if 'vibrant' in tools:
        df['vb_lifestyle'].fillna('Undetermined', inplace=True)
        df['vb_score'] = df['vb_isPhage'].apply(lambda x : 1 if x == 1 else 0)
        score += df['vb_score']
    if 'deepvirfinder' in tools:
        df['dvf_score'] = df.apply(lambda x : 1 if x['dvf_v_score'] >= 0.9 and x['dvf_pvalue'] <= 0.05 else 0, axis = 1)
        score += df['dvf_score']
    if 'genomad' in tools:
        df['gn_taxonomy'].fillna('Unclassified', inplace=True)
        df['gn_score'] = df.apply(lambda x : 2 if x['gn_v_score'] >= 0.8 else (1 if x['gn_v_score'] >= 0.7 and x['gn_hallmarks'] >= 1 else -1), axis = 1)
        score += df['gn_score']
    df['score'] = score
    postfix = '.score.tsv'
    all_ctg_score_tsv = all_merged_ctg_tsv.replace('.tsv', postfix)
    df.to_csv(all_ctg_score_tsv, index=False, sep='\t')
    score_cutoff_dict = {'permissive': 1, 'strict': 2}
    postfix = f'.score.{filt_mode}.tsv'
    all_ctg_score_filt_tsv = all_merged_ctg_tsv.replace('.tsv', postfix)
    df = df.query(f'score >= {score_cutoff_dict[filt_mode]}')
    df.to_csv(all_ctg_score_filt_tsv, index=False, sep='\t')
    return 0

def ctgList(name, wkdir, tools=['gn'], filt_mode='permissive'):
    all_nh_ctgs = vCtgMerge(name, wkdir, tools)
    all_ctgs = ['Contig'] + all_nh_ctgs
    all_ctgs_li = f'{wkdir}/all_viral_ctgs.list'
    listToFile(all_ctgs, all_ctgs_li)
    mark = all_ctgs_li
    for tool in tools:
        csv_name = CsvDict[tool].format(wkdir)
        merged_name = mark + '_' + tool
        linkTab.merge(mark, csv_name, 'left', 'Contig', merged_name)
        if mark != all_ctgs_li:
            os.remove(mark)
        mark = merged_name
    all_merged_ctg_tsv = f'{wkdir}/all_viral_ctgs.tsv'
    os.rename(mark, all_merged_ctg_tsv)
    rcode = calcCtgScore(all_merged_ctg_tsv, tools, filt_mode)
    return rcode

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('name')
    parser.add_argument('wkdir')
    parser.add_argument('-m', '--tool-chain', default='gn')
    parser.add_argument('-f', '--filter-mode', default='permissive', choices=['permissive', 'strict'])
    args = parser.parse_args()
    tools = parse_merge_tools(args.tool_chain)
    ctgList(args.name, args.wkdir, tools, args.filter_mode)
